c++/src/capnp/compiler/capnpc-c++.c++

```diff
@@ -2188,7 +2188,12 @@ private:
   InterfaceText makeInterfaceText(kj::StringPtr scope, kj::StringPtr name, InterfaceSchema schema,
                                   kj::Array<kj::StringTree> nestedTypeDecls,
                                   const TemplateContext& templateContext) {
-    auto fullName = kj::str(scope, name, templateContext.args());
+    auto templateArgs = templateContext.args();
+    auto fullName = kj::str(scope, name, templateArgs);
+    auto fullDisambiguatedName = kj::str(
+        scope.size() ? "typename " : "", scope,
+        scope.size() && templateArgs.size() ? "template " : "", name,
```

Comment:
Hmm, is there a problem here if one of the parent scopes (encoded in `scope`) is a template? It looks like you're only adding `template` to the last component.

Actually, I wonder if this stuff is even necessary. Looking at the code for structs, the `Reads` and `Builds` typedefs there are defined using the unqualified parent type, i.e.:

    struct MyType {
      class Reader {
        typedef MyType Reads;

I'm not sure why the `Calls`, `Reads`, and `Serves` typedefs of interfaces are using fully-qualified names here -- maybe they don't have to?